src/hardware/configuration.rs

```diff
@@ -105,6 +105,32 @@ pub struct PounderDevices {
 /// Static storage for the ethernet DMA descriptor ring.
 static mut DES_RING: ethernet::DesRing = ethernet::DesRing::new();
 
+/// Setup ITCM and load its code from flash
+unsafe fn setup_itcm() {
```

Comment:
Can we just tag this as `#[cortex_m_rt::pre_init]` and avoid having to manually call it during setup? Is there a reason we're doing it later on after hardware initialization (e.g. does this require RCC config, etc)? If so, we should document.

Reply:
As mentioned, this can't be pre_init because for example the stack isn't set up. It can be earlier in setup() but there is no real point in doing that. Manually calling it seems fine. We are calling pretty much everything manually.

Reply:
The stack is set up before `Reset` is even called - this is done by the ARM architecture - the SP is configured as the first entry in the vector table and is part of the CPU boot process:
https://github.com/rust-embedded/cortex-m-rt/blob/master/link.x.in#L63

If the stack wasn't set up, one could _never_ use the pre-init hook.

At this stage, the only thing not initialized is the FPU and BSS/DATA. Given that we only use a single constant (which is RODATA, stored in flash), we should be initialized enough to execute this function in the pre-init hook (this is what the pre-init hook is designed for).

The reason I'm pushing this is that if we don't do this as early as possible during startup, we are leaving a foot-gun around for ourselves later. If we (for some reason) put some init code in the `ITCM` in the future and forget exactly where the ITCM is initialized, we might kill hours debugging.

Did you try running this as a `pre_init()` hook already and observe that it didn't work?

Reply:
These are good points - I would encourage you to add the discussion here (relevant points thereof) as  documentation to this initialization function